src/Dispatch.elm

```diff
@@ -0,0 +1,262 @@
+module Dispatch
+  exposing
+    ( Msg
+    , Decoder
+    , forward
+    , listeners
+    , listeners'
+    , group
+    , update
+    )
+
+{-| Utility module for dispatching multiple events from a single `Html.Event`
+
+## Types
+@docs Msg
+@docs Decoder
+
+## Event handlers
+@docs listeners
+@docs listeners'
+@docs group
+
+## Dispatch
+@docs update
+@docs forward
+-}
+
+import Json.Decode as Json
+import Html.Events
+import Html.Attributes
+import Html
+import Task
+
+
+
+{-|
+  Lift any value of type `msg` to a `Cmd msg`.
+-}
+cmd : msg -> Cmd msg
+cmd msg =
+  Task.perform (always msg) (always msg) (Task.succeed msg)
+
+
+
+{-| NoOp
+ -}
+noop : Html.Attribute msg
+noop = Html.Attributes.attribute "data-dispatch-noop" "noop"
+
+{-| Message type
+-}
+type alias Msg m
+  = (List m)
+
+
+{-| A decoder with possible options
+-}
+type alias Decoder m =
+  ( Json.Decoder m, Maybe (Html.Events.Options) )
+
+
+{-| Maps messages to commands
+-}
+forward : Msg m -> Cmd m
+forward (messages) =
+  List.map cmd messages |> Cmd.batch
+
+
+{-| Map the second element of a tuple
+
+    map2nd ((+) 1) ("bar", 3) == ("bar", 4)
+-}
+map2nd : (b -> c) -> ( a, b ) -> ( a, c )
+map2nd f ( x, y ) =
+  ( x, f y )
+
+
+{-| Runs batch update
+-}
+update : (a -> b -> ( b, Cmd c )) -> Msg a -> b -> ( b, Cmd c )
+update update (msg) model =
+  let
+    inner cmd ( m, gs ) =
+      update cmd m
+        |> map2nd (flip (::) gs)
+  in
+    List.foldl
+      inner
+      ( model, [] )
+      msg
+      |> map2nd Cmd.batch
+
+
+{-| Applies given decoders to the same initial value
+   and return the applied results as a list
+-}
+applyMultipleDecoders : List (Json.Decoder m) -> Json.Decoder (List m)
+applyMultipleDecoders decoders =
+  let
+    processDecoder initial decoder =
+      Json.decodeValue decoder initial
+        |> Result.toMaybe
+  in
+    Json.customDecoder Json.value
+      (\initial ->
```

Comment:
Lift the closure out of applyMultipleDecoders. (Won't affect the benchmark, which never tries multiple-event dispatching.)